Added first revision of code

gcoreddy opened the pull request:
Added first revision of code to the new feature branch

gcoreddy committed:
Added first revision of code

Added first revision of code to the new feature branch

kingdomik commented:
Hi  @gcoreddy patch contains *,.pyc files

kingdomik commented on parking/migrations/0001_initial.py:
```diff
@@ -0,0 +1,41 @@
+# Generated by Django 3.0.5 on 2020-04-09 16:24
```
Do we really need to place generated files to version control?

kingdomik commented on README.md:
```diff
@@ -0,0 +1,2 @@
+# parkingStationRestWeb
+ parking station implementation in rest as well as web
```
Having such documentation we can't present it to customer :) how to use it?

gcoreddy replied:
I have added more read me in latest changeset.

kingdomik commented on parking/src/parking.py:
```diff
@@ -0,0 +1,172 @@
+            # Exception if there is multiple cars allocated with the same location.
+            receipt = parkingExceptionsDict["MultipleCarsWithSameLocation"]
+        finally:
+            return receipt
```
Cool idea!

kingdomik commented on parking/views.py:
```diff
@@ -0,0 +1,53 @@
+            car_num = request.GET['car_num']
+            tariff_plan = request.GET['tariff_plan']
+        except:
+            responseDict = {"status" : "error", "code" : 400, "Reason": "<car_num> and <tariff_plan> should be provided."}
```
I believe exception processing expects specific exception but not any one

kingdomik commented on parking/src/parking.py:
```diff
@@ -0,0 +1,172 @@
+            free_spots = list(self.jsonDec.decode(level.free_spots))
+            if len(free_spots) > 0:
+                # Assign a parking spot in this specific level if free spots available.
+                assignedSpot = free_spots[-1]
```
Need to explain in comment why do we choose latest slot

kingdomik commented on parking/src/parking.py:
```diff
@@ -0,0 +1,172 @@
+            if len(free_spots) > 0:
+                # Assign a parking spot in this specific level if free spots available.
+                assignedSpot = free_spots[-1]
+                location = level.level_num + "_" + str(assignedSpot)
```
Why two different data were concatenated? Data representation shouldn't influence on core code since representation can be changed later and core should stay

kingdomik commented on parking/src/parking.py:
```diff
@@ -0,0 +1,172 @@
+                location = level.level_num + "_" + str(assignedSpot)
+                break
+        # Raise an exception if no space left in any of the parking level.
+        if location is None: raise NoSpaceLeftInParking
```
According PEP8 each statement must be on a separate line

kingdomik commented on parking/src/parking.py:
```diff
@@ -0,0 +1,172 @@
+            pass
+        else:
+            # Raise an exception if car already there with the same name.
+            raise CarWithSameNumExist
```
Method requests car but never use it, why do we need this function?

kingdomik replied:
Got idea

kingdomik commented on parking/src/parking.py:
```diff
@@ -0,0 +1,172 @@
+            receipt = parkingExceptionsDict["CarWithSameNumExist"]
+        except TariffPlan.DoesNotExist:
+            # Exception if there is no tariff plan defined.
+            receipt = parkingExceptionsDict["TariffPlanDoesntExist"]
```
Error doesn't contain enough information to investigate failure

kingdomik commented on parking/src/common.py:
```diff
@@ -0,0 +1,85 @@
+             "NoSpaceLeftInParking"  : { "status" : "Error",
+			                             "code"   : 404,
+			                              "Reason" : "No Space left in the parking."
+							           },
```
I like that all error messages are collected in same place!
But can we integrate messages and codes directly with exceptions?

kingdomik commented on parking/src/parking.py:
```diff
@@ -0,0 +1,172 @@
+            receipt = parkingExceptionsDict["LocationEmpty"]
+        except carDataDetails.MultipleObjectsReturned:
+            # Exception if there is multiple cars allocated with the same location.
+            receipt = parkingExceptionsDict["MultipleCarsWithSameLocation"]
```
What about processing of unexpected exceptions?

kingdomik commented on parking/src/parking.py:
```diff
@@ -0,0 +1,172 @@
+            # Retrieve and return each car details parked in the parking station.
+            car["inTime"] = time.strftime("%m/%d/%Y, %H:%M:%S",time.gmtime(float(car["inTime"])))
+            carDetailsDict["cars"].append(car)
+        carDetailsDict['code'] = 204 if len(carDetailsDict['cars']) == 0 else 200
```
REST specific shouldn't be hardcoded in core engine

kingdomik commented on parking/src/Ticket.py:
```diff
@@ -0,0 +1,86 @@
+        """Prints the ticket at entry and exit also calculates the fare as well at exit."""
+        ticketDict = { "Car" : self.carno, "tariff" : self.tariff , 
+                       "Location": self.location, "ID": self.ID,
+                       "Start" : time.strftime("%m/%d/%Y, %H:%M:%S",time.gmtime(float(self.inTime)))}
```
Time conversion duplicated in many places

kingdomik commented on parking/src/Ticket.py:
```diff
@@ -0,0 +1,86 @@
+        for plan in TariffPlan.objects.all().values("plan","cost","freetime"):
+            tarifPlansDict["TariffPlansInfo"].append(plan)
+        tarifPlansDict["status"] = "Success"
+        tarifPlansDict["code"] = 200
```
Representation shouldn't import core code

kingdomik commented on parking/src/parking.py:
```diff
@@ -0,0 +1,172 @@
+		   in the specified level.
+		'''
+        # Retrieve level object.
+        level = ParkingLevel.objects.get(level_num=location.split("_")[0])
```
location is combined from two values before now split, don't think it is a good idea

kingdomik commented on parking/src/parking.py:
```diff
@@ -0,0 +1,172 @@
+        assignedSpot = location.split("_")[1]
+        # Remove a assigned spot from free spots and update the db.
+        free_spots.remove(int(assignedSpot))
+        level.free_spots = json.dumps(list(free_spots))
```
Looks strange to decode/encode JSON, seems some mistake in design

kingdomik commented on parking/src/Ticket.py:
```diff
@@ -0,0 +1,86 @@
+            tariffMnts = self.tariffMinutesDict[self.tariff]
+            totalTime  = self.outTime - self.inTime - float(freetime)*60
+            price = (totalTime//int(tariffMnts)+1) * int(cost)
+            return price
```
IMPORTANT: We need more flexible model of tariff plan since customer is going to use our code for different with absolutely different plans and our core code should be able to easily support any algorithm.

gcoreddy replied:
Added new alogorithm to calculate the cost.
Please refer the latest readme for the details.

gcoreddy committed:
Added readme

kingdomik commented:
Hi @gcoreddy could you please ping me when code be ready for next review. I see something was fixed but issues don't yet

gcoreddy committed:
Removed all the binary files.

Removed all the pyc and migration files.
User has to run makemigrations before starting the app.

gcoreddy committed:
Merge branch 'featurebranch' of https://github.com/gcoreddy/ParkingStationProject into featurebranch

gcoreddy commented:
> Hi @gcoreddy could you please ping me when code be ready for next review. I see something was fixed but issues don't yet

Removed all the pyc files from the review.

gcoreddy commented:
> Hi @gcoreddy could you please ping me when code be ready for next review. I see something was fixed but issues don't yet

I somehow missed the notification in my gmail. I just saw your comment today only.

gcoreddy committed:
Updated readme file.

Formatted the readme page.

carlosheiras17 commented:
Hello, I would like to ask what's the style guide adopted for this project?

carlosheiras17 commented on parking/views.py:
```diff
@@ -0,0 +1,53 @@
+    if request.method == 'GET':
+        return render(request, 'index.html', {})
+    elif request.method == 'POST':
+        if request.POST['OPTION'] == "1":
```
what does 1,2,3,4,5 means? please avoid magic numbers or similar.

carlosheiras17 commented on parking/src/parking.py:
```diff
@@ -0,0 +1,172 @@
+	   also for extending the Levels, updating tariff plans'''
+    def __init__(self):
+        '''Constructor method'''
+        self.jsonDec = json.decoder.JSONDecoder()
```
let's inject decoder into Class to make it more flexible.

carlosheiras17 commented on parking/src/parking.py:
```diff
@@ -0,0 +1,172 @@
+        
+    def addLevel(self,level_name,total_spots):
+        '''Add or extend a new parking llevel to the station.'''
+        if len(ParkingLevel.objects.filter(level_num=level_name)) > 0:
```
what does `ParkingLevel.objects.filter(level_num=level_name)` returns? a list?

carlosheiras17 commented on parking/src/parking.py:
```diff
@@ -0,0 +1,172 @@
+            # Create Ticket with fare details.
+            ticket = Ticket(location)
+            receipt = ticket.printTicket("Entry")
+        except ParkingLevelDoesntExist:
```
I like how error messages for parking lot are in only one place, but maybe we can work with inheritance somehow, to achieve a clean `try...except`, I mean, the only thing changing from line 92 to 103 are custom exceptions (`except CustomException`), as well as custom exception name as a string (`receipt = parkingExceptionsDict["CustomException"]`).

carlosheiras17 commented on parking/models.py:
```diff
@@ -0,0 +1,21 @@
+from __future__ import unicode_literals
```
in general for this file:
why attributes that are supposed to be manipulated as digits/integers/numbers like cost, freetime, inTime are declared as ChardField?

carlosheiras17 replied:
also, if this is core file for ORM, why these models/tables aren't relationed by foreign_keys?

carlosheiras17 commented on parking/src/Ticket.py:
```diff
@@ -0,0 +1,86 @@
+            price = (totalTime//int(tariffMnts)+1) * int(cost)
+            return price
+			
+    def createOrUpdateTariff(plan_name,plan_cost,plan_freetime):
```
please apply single responsability principle to this function.

carlosheiras17 reviewed (changes requested):
before continuing my review, please let me know if current implementation is Model-View-Controller or Model-view-viewmodel.

carlosheiras17 commented on parking/src/Ticket.py:
```diff
@@ -0,0 +1,86 @@
+    def __init__(self, location) :
+        '''Constructor method'''
+        self.outTime = time.time()
+        car = self.__findVehical(location)
```
don't load data inside object constructors, please inject any complex structure so this way if creation of mentioned complex data/structure raises an error it doesn't belong to class instantiation

carlosheiras17 commented on parking/src/Ticket.py:
```diff
@@ -0,0 +1,86 @@
+        tariff = TariffPlan.objects.get(plan=self.tariff)
+        return (tariff.cost,tariff.freetime)
+
+    def __calAmount(self):
```
don't hesitate to express want you want in function names, `def _calculate_amount()` its a good idea. 

_ as very personal opinion I think long names in private methods aren't that bad, while short and expressive names are very good for interfaces/public methods._

carlosheiras17 commented on parking/src/parking.py:
```diff
@@ -0,0 +1,172 @@
+        try:
+            # Retrieving the car details with the name.
+            car = carDataDetails.objects.get(carno=car_num)
+        except carDataDetails.DoesNotExist:
```
between not catching `carDataDetails.DoesNotExis` and `except carDataDetails.DoesNotExist: pass`, the second one is the worst option, please take care of this anti-pattern.

gcoreddy committed:
Addressed comments of Carlos.

Updated readme.
Added Foriegn keys.
Imporved the exception handling.
Addressed all comments.

gcoreddy committed:
Update README.md

gcoreddy committed:
Confilcts resolved

Resolved conflicts in Readme

gcoreddy commented:
> Hello, I would like to ask what's the style guide adopted for this project?
I have addressed all the comments.
Added the Architecture i have used in the Read me file.

gcoreddy committed:
Modified the code as per pep8  style guide

Modified the code as per the pep8 python standard style guide.

gcoreddy commented:
> Hello, I would like to ask what's the style guide adopted for this project?

I have adopted PEP8 style guide for this. Latest commit code modified as per this style guide.

gcoreddy committed:
Addressed comments of Denis.

Added new algorithm which calculates the time based on total time and suggests the user the better plan.

gcoreddy committed:
Update README.md
